File: apple/Packages/DsmCore/Sources/NasAddress.swift
import Foundation

public enum NasAddressKind: Equatable, Sendable {
    case direct
    case quickConnect
}

public struct ParsedNasAddress: Equatable, Sendable {
    public let host: String
    public let port: Int
    public let kind: NasAddressKind
    public let hasExplicitPort: Bool

    public init(
        host: String,
        port: Int,
        kind: NasAddressKind,
        hasExplicitPort: Bool = false
    ) {
        self.host = host
        self.port = port
        self.kind = kind
        self.hasExplicitPort = hasExplicitPort
    }
}

public enum NasAddressInputError: Error, Equatable, Sendable {
    case empty
    case invalid
    case insecure
}

extension NasAddressInputError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .empty:
            "请输入 NAS 地址或 QuickConnect ID。"
        case .invalid:
            "无法识别这个地址。你可以粘贴浏览器中的完整地址，或输入 IP、域名和 QuickConnect ID。"
        case .insecure:
            "为了保护登录信息，请使用 HTTPS 地址。"
        }
    }
}

public enum NasAddressParser {
    public static func parse(_ input: String, defaultPort: Int) throws -> ParsedNasAddress {
        guard (1...65_535).contains(defaultPort) else {
            throw NasAddressInputError.invalid
        }

        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw NasAddressInputError.empty
        }

        let hasExplicitScheme = trimmed.contains("://")
        let value = hasExplicitScheme ? trimmed : "https://\(trimmed)"
        guard let components = URLComponents(string: value),
              let rawHost = components.host,
              !rawHost.isEmpty,
              components.user == nil,
              components.password == nil else {
            throw NasAddressInputError.invalid
        }

        let host = rawHost.lowercased()
        if let quickConnectID = quickConnectID(host: host, path: components.path) {
            guard isPotentialQuickConnectID(quickConnectID) else {
                throw NasAddressInputError.invalid
            }
            return ParsedNasAddress(
                host: quickConnectID,
                port: 5_001,
                kind: .quickConnect,
                hasExplicitPort: false
            )
        }

        guard components.scheme?.lowercased() == "https" else {
            throw NasAddressInputError.insecure
        }

        let port = components.port ?? (hasExplicitScheme ? 443 : defaultPort)
        guard (1...65_535).contains(port) else {
            throw NasAddressInputError.invalid
        }

        return ParsedNasAddress(
            host: host,
            port: port,
            kind: isPotentialQuickConnectID(host) ? .quickConnect : .direct,
            hasExplicitPort: components.port != nil
        )
    }

    public static func isPotentialQuickConnectID(_ value: String) -> Bool {
        guard (1...64).contains(value.count),
              !value.contains("."),
              !value.contains(":"),
              value.first != "-",
              value.last != "-" else {
            return false
        }
        let allowed = CharacterSet(
            charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-"
        )
        return value.unicodeScalars.allSatisfy(allowed.contains)
    }

    private static func quickConnectID(host: String, path: String) -> String? {
        let portalHosts = ["quickconnect.cn", "quickconnect.to"]
        if portalHosts.contains(host) {
            return path.split(separator: "/").first.map(String.init)?.lowercased()
        }

        for suffix in portalHosts {
            let expectedSuffix = ".\(suffix)"
            if host.hasSuffix(expectedSuffix),
               host.split(separator: ".").count == 3 {
                return String(host.dropLast(expectedSuffix.count)).lowercased()
            }
        }
        return nil
    }
}

File: android/app/src/main/java/io/github/qwertyuiop1995/dsmnativeclient/network/NasAddress.kt
package io.github.qwertyuiop1995.dsmnativeclient.network

import io.github.qwertyuiop1995.dsmnativeclient.domain.DsmFailure
import io.github.qwertyuiop1995.dsmnativeclient.domain.DsmErrorKind
import java.net.URI

internal enum class NasAddressKind {
    DIRECT,
    QUICK_CONNECT,
}

internal data class ParsedNasAddress(
    val host: String,
    val port: Int,
    val kind: NasAddressKind,
    val hasExplicitPort: Boolean,
)

/**
 * 解析用户保存的 NAS 地址。
 *
 * QuickConnect ID 只在连接期间解析为临时候选地址，不能覆盖用户保存的原始 ID。
 */
internal object NasAddressParser {
    fun parse(input: String, portOverride: Int? = null): ParsedNasAddress {
        if (portOverride != null && portOverride !in 1..65_535) {
            throw invalidAddress()
        }
        val trimmed = input.trim()
        if (trimmed.isEmpty()) {
            throw DsmFailure(
                null,
                "NAS address or QuickConnect ID is required",
                "Enter the address and connect again.",
                kind = DsmErrorKind.INVALID_ADDRESS,
            )
        }

        val hasExplicitScheme = "://" in trimmed
        val uri = runCatching {
            URI(if (hasExplicitScheme) trimmed else "https://$trimmed")
        }.getOrElse {
            throw invalidAddress()
        }
        if (uri.userInfo != null || uri.fragment != null || uri.query != null) {
            throw invalidAddress()
        }

        val host = uri.host?.lowercase()?.takeIf(String::isNotBlank)
            ?: throw invalidAddress()
        quickConnectId(host, uri.path.orEmpty())?.let { id ->
            if (!isPotentialQuickConnectId(id)) throw invalidAddress()
            return ParsedNasAddress(
                host = id,
                port = portOverride ?: DEFAULT_DSM_HTTPS_PORT,
                kind = NasAddressKind.QUICK_CONNECT,
                hasExplicitPort = portOverride != null,
            )
        }

        if (uri.scheme?.lowercase() != "https") {
            throw DsmFailure(
                null,
                "The address is not secure",
                "Use an HTTPS address to protect sign-in information.",
                kind = DsmErrorKind.INSECURE_ADDRESS,
            )
        }
        if (isPotentialQuickConnectId(host)) {
            return ParsedNasAddress(
                host = host,
                port = portOverride ?: DEFAULT_DSM_HTTPS_PORT,
                kind = NasAddressKind.QUICK_CONNECT,
                hasExplicitPort = portOverride != null,
            )
        }

        val explicitUriPort = uri.port.takeIf { it > 0 }
        val port = portOverride
            ?: explicitUriPort
            ?: if (hasExplicitScheme) DEFAULT_HTTPS_PORT else DEFAULT_DSM_HTTPS_PORT
        if (port !in 1..65_535) throw invalidAddress()
        return ParsedNasAddress(
            host = host,
            port = port,
            kind = NasAddressKind.DIRECT,
            hasExplicitPort = portOverride != null || explicitUriPort != null,
        )
    }

    fun isPotentialQuickConnectId(value: String): Boolean {
        if (value.length !in 1..64 ||
            '.' in value ||
            ':' in value ||
            value.startsWith('-') ||
            value.endsWith('-')
        ) {
            return false
        }
        return value.all(::isAsciiHostCharacter)
    }

    private fun quickConnectId(host: String, path: String): String? {
        if (host in QUICK_CONNECT_PORTALS) {
            return path.split('/').firstOrNull(String::isNotBlank)?.lowercase()
        }
        return QUICK_CONNECT_PORTALS.firstNotNullOfOrNull { portal ->
            val suffix = ".$portal"
            host.takeIf { it.endsWith(suffix) && it.count { character -> character == '.' } == 2 }
                ?.removeSuffix(suffix)
                ?.lowercase()
        }
    }

    private fun invalidAddress() = DsmFailure(
        null,
        "The address is invalid",
        "Enter a QuickConnect ID, NAS IP address, domain, or full HTTPS address.",
        kind = DsmErrorKind.INVALID_ADDRESS,
    )

    private const val DEFAULT_HTTPS_PORT = 443
    private const val DEFAULT_DSM_HTTPS_PORT = 5_001
    private val QUICK_CONNECT_PORTALS = setOf("quickconnect.to", "quickconnect.cn")

    private fun isAsciiHostCharacter(value: Char): Boolean =
        value in 'a'..'z' || value in 'A'..'Z' || value in '0'..'9' || value == '-'
}
